feat: create feature request api endpoint

DanielleBadobre opened the pull request:
## FEATURE REQUESTS API

## Description

Implemented feature request functionality, including:
- Feature request API endpoints (Create, Read, Update, Delete)
- Authorization checks to allow only request owners or superadmins to access and modify requests
- Pydantic request and response models
- Database service layer integration
- Unit tests for API endpoints

## Related Issue (Link to issue ticket)
https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/1191

## Motivation and Context

This feature enables users to request new functionalities directly from the application while ensuring that only authorized users can view, update, or delete their own requests. Superadmins have access to all feature requests.

## How Has This Been Tested?

- Postman: Manually tested all API endpoints with valid and invalid requests
- FastAPI documentation UI: Verified request creation, retrieval, update, and deletion
- Pytest: Wrote unit tests to cover major service layer functions and API endpoints

## Screenshots (Postman, etc):

#### Pytest Results

![pytests](https://github.com/user-attachments/assets/fff625d8-0b55-4312-ad33-00b9b9cd9c84)

#### Creating Feature Request

![create feature](https://github.com/user-attachments/assets/655b18a0-0ff3-4832-8785-b07a0ee60d0b)

#### Getting User's Features Requests

![user features](https://github.com/user-attachments/assets/8b92ef5a-512f-450f-a306-4e016e8f0c56)

#### Getting Feature Request by ID

![Ifeature by id](https://github.com/user-attachments/assets/f3e47c45-92bd-470d-ba88-3a2b160c5d0f)

#### Updating Feature Request

![updated by id](https://github.com/user-attachments/assets/0155f77a-1b96-4293-808f-f421fb3fdd24)

#### Deleting Feature Request

![deleting feature](https://github.com/user-attachments/assets/667ff090-07ec-4a96-9287-e10ca1a0677f)

#### Accessing Deleted Feature Request

![deleted requests](https://github.com/user-attachments/assets/f9b70c49-37c6-4270-88fe-9dda74c20374)
## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

DanielleBadobre committed:
feat: implemented feature request api

DanielleBadobre committed:
Merge branch 'dev' of https://github.com/DanielleBadobre/hng_boilerplate_python_fastapi_web into dev
feat: implement feature requests api

DanielleBadobre committed:
Fix(test): feature request tests with correct priority type and mock paths

DanielleBadobre committed:
fix(tests): Resolve test failures and improve test coverage for feature request endpoints

- Fixed mock setups for `create_feature_request`, `get_feature_requests`, and related functions to ensure proper calls.
- Reordered authorization checks to prioritize "not found" checks before "forbidden" checks.
- Corrected mock mismatches in `get_feature_request_success_admin` and other tests.
- Improved test assertions and coverage for error cases (404, 403) and success scenarios.

DanielleBadobre committed:
Merge branch 'dev' into dev

joboy-dev commented on alembic.ini:
remove this change

DanielleBadobre replied:
done ✅

joboy-dev reviewed (changes requested):
Remove all these pictures i am seeing here
There are other ways to upload a picture into your PR

DanielleBadobre committed:
chore: remove sqlalchemy test url from alembic.ini

Signed-off-by: DanielleBadobre <109845369+DanielleBadobre@users.noreply.github.com>

DanielleBadobre committed:
chore: Delete tests/v1/feature_request/screenshots directory

Signed-off-by: DanielleBadobre <109845369+DanielleBadobre@users.noreply.github.com>

DanielleBadobre committed:
Merge branch 'dev' into dev

DanielleBadobre commented:
> Remove all these pictures i am seeing here There are other ways to upload a picture into your PR

done ✅

samuelogboye commented on api/v1/services/feature_request.py:
```diff
@@ -0,0 +1,77 @@
+    title: str = Field(..., description="Title of the feature request")
+    description: str = Field(..., description="Detailed description of the requested feature")
+    priority: str = Field(default="Low", description="Priority level (Low, Medium, High)")
+    status: str = Field(default="Pending", description="Status (Pending, Approved, Rejected)")
+
+
+class FeatureRequestCreate(FeatureRequestBase):
+    pass
+
+
+class FeatureRequestUpdate(BaseModel):
+    title: Optional[str] = Field(None, description="Title of the feature request")
+    description: Optional[str] = Field(None, description="Detailed description of the requested feature")
+    priority: Optional[str] = Field(None, description="Priority level (Low, Medium, High)")
+    status: Optional[str] = Field(None, description="Status (Pending, Approved, Rejected)")
```
remove status from required input during creation

Appending the status should be solely done by the server.

The server should ALWAYS give an initial pending status

which then can be updated by the admin later on

DanielleBadobre replied:
Updated the status field to always be pending until an admin changes it and created tests accordingly ✅

DanielleBadobre committed:
Merge branch 'hngprojects:dev' into dev

DanielleBadobre committed:
refactor: remove status from FeatureRequestCreate schema

Signed-off-by: DanielleBadobre <109845369+DanielleBadobre@users.noreply.github.com>

DanielleBadobre committed:
feat: restrict feature request status updates to admin users only

Signed-off-by: DanielleBadobre <109845369+DanielleBadobre@users.noreply.github.com>

DanielleBadobre committed:
feat: Set feature request status to "Pending" and remove duplicate schemas

Signed-off-by: DanielleBadobre <109845369+DanielleBadobre@users.noreply.github.com>

DanielleBadobre committed:
test: update feature request tests for status restriction

- Modify fixtures to remove status from feature request creation data
- Add tests to verify status is always set to "Pending" on creation
- Add test cases covering admin-only status update restriction
- Ensure non-admin users can still update other fields
- Fix capitalization of priority values to match model defaults

Signed-off-by: DanielleBadobre <109845369+DanielleBadobre@users.noreply.github.com>

DanielleBadobre committed:
Merge branch 'hngprojects:dev' into dev

DanielleBadobre committed:
Merge branch 'dev' into dev

samuelogboye commented on api/v1/routes/feature_request.py:
```diff
@@ -0,0 +1,134 @@
+
+     # Prevent non-superadmins from updating the status field
+    update_data = feature_request_update.dict(exclude_unset=True)
+    if not current_user.is_superadmin and "status" in update_data:
+        raise HTTPException(
+            status_code=status.HTTP_403_FORBIDDEN,
+            detail="Only admins can update the status field"
+        )
```
Youve already made a check to verify admin access in line 89
these lines of code will never be reached.

You can remove it

DanielleBadobre replied:
ok sir, but actually the two checks serve different purposes. 

- The first check (line 89) verifies if the user has permission to update the feature request at all. It checks if the user is either a superadmin OR the owner of the feature request.

- The second check (lines 98-102) only applies when a user is trying to update the "status" field. Even if a user passes the first check (because they're the owner of the request), they'll still hit this second check if they're trying to update the status field. 

 should i still remove the lines ?

samuelogboye reviewed (approved):
Resolve that change and we are GTG
